File: python/gentrail/enforcement.py
"""Pre-call policy enforcement.

The async evaluator only sees a trace after the tool already ran, so it can
detect but never prevent. Enforcement therefore happens here, in the SDK, at the
before-tool-call hook: ask the backend for a verdict on the proposed tool call
and cancel it on BLOCK before it executes.

Opt-in: a PolicyEnforcer is built only when GENTRAIL_DECIDE_ENDPOINT and
GENTRAIL_API_KEY are both set, so the default behaviour stays observe-only.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import time
import urllib.request
import uuid

logger = logging.getLogger(__name__)

GATE_POLL_INTERVAL_SECONDS = 2.0

# Identify the client on every request. urllib's default "Python-urllib/x.y"
# reads as a bot to edge WAFs (e.g. Cloudflare fronting the backend answers 403),
# which would silently fail-open enforcement; a named agent is allowed through.
_USER_AGENT = "gentrail-sdk-python"


class PolicyEnforcer:
    """Synchronous client for the backend's /api/v1/decide endpoint."""

    def __init__(self, endpoint: str, api_key: str, timeout: float = 3.0):
        self.base = endpoint.rstrip("/")
        self.url = self.base + "/api/v1/decide"
        self.api_key = api_key
        self.timeout = timeout
        # A GATE hold blocks the tool while a human decides. Bound the wait so an
        # unanswered gate never hangs the agent forever; on timeout the gate
        # fails closed (the tool stays cancelled). The backend hold expires
        # independently (deny-by-default) at its own expires_at.
        self.gate_timeout = float(
            os.environ.get("GENTRAIL_GATE_TIMEOUT_SECONDS", "120")
        )

    @classmethod
    def from_env(cls) -> "PolicyEnforcer | None":
        endpoint = os.environ.get("GENTRAIL_DECIDE_ENDPOINT", "").strip()
        api_key = os.environ.get("GENTRAIL_API_KEY", "").strip()
        if not endpoint or not api_key:
            return None
        return cls(endpoint, api_key)

    def decide(
        self,
        tool_name: str,
        tool_args: dict,
        *,
        agent_id: str = "",
        invocation_id: str = "",
        request_id: str = "",
    ) -> dict:
        """Return the backend verdict: {"decision": BLOCK|GATE|ALLOW, "rule", "message"}.

        agent_id scopes agent-targeted and windowed rules to the caller;
        invocation_id (the invocation's OTel trace id) lets the backend join
        the enforcement record to the ingested trace and gives a GATE approver
        context, and defaults to the ambient trace when a span is recording;
        request_id makes the backend's GATE/BLOCK writes idempotent under
        retries. The backend requires a request_id, so a caller without a
        framework call id gets a synthesized one (unique, so it never dedups a
        legitimate second call). It refuses to record a BLOCK/GATE with no
        agent_id or invocation_id, since it could not be joined to its trace
        and would double-count against the evaluator's own row.

        Fails open - a backend error must never break the agent, only forgo
        enforcement for that call.
        """
        payload = {
            "event_type": "tool_call",
            "tool_name": tool_name,
            "tool_args": tool_args,
            "request_id": request_id or uuid.uuid4().hex,
        }
        if agent_id:
            payload["agent_id"] = agent_id
        invocation_id = invocation_id or ambient_otel_trace_id()
        if invocation_id:
            payload["invocation_id"] = invocation_id
        body = json.dumps(payload).encode()
        req = urllib.request.Request(
            self.url,
            data=body,
            method="POST",
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self.api_key}",
                "User-Agent": _USER_AGENT,
            },
        )
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode())
        except Exception as e:
            logger.warning("enforcement decide failed (%s); allowing tool %s", e, tool_name)
            return {"decision": "ALLOW"}

    def await_gate(self, approval: dict, *, timeout: float | None = None) -> str:
        """Poll a GATE hold's status resource until it resolves, returning the
        final status: "approved", "denied", "expired", or "timeout".

        Unlike decide(), this fails CLOSED: an unresolved or unreachable hold
        returns a non-approved status so the caller keeps the tool cancelled.
        Running a gated action without a confirmed approval is exactly what the
        gate exists to prevent.
        """
        status_url = (approval or {}).get("status_url")
        if not status_url:
            return "timeout"
        poll_url = self.base + status_url
        deadline = time.monotonic() + (self.gate_timeout if timeout is None else timeout)
        while True:
            try:
                status = self._poll_gate_once(poll_url)
            except Exception as e:
                logger.warning("gate poll failed (%s); holding the gate closed", e)
                return "timeout"
            if status != "pending":
                return status
            if time.monotonic() >= deadline:
                return "timeout"
            time.sleep(GATE_POLL_INTERVAL_SECONDS)

    def _poll_gate_once(self, poll_url: str) -> str:
        req = urllib.request.Request(
            poll_url,
            method="GET",
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "User-Agent": _USER_AGENT,
            },
        )
        with urllib.request.urlopen(req, timeout=self.timeout) as resp:
            return json.loads(resp.read().decode()).get("status", "pending")


class AsyncPolicyEnforcer:
    """Async twin of PolicyEnforcer for asyncio-native callers.

    Each HTTP round trip is the sync urllib call moved off the event loop with
    asyncio.to_thread, so the base install still needs no HTTP client
    dependency, while GATE polling waits with asyncio.sleep instead of
    parking a thread. Semantics match the sync client exactly: decide fails
    open, await_gate fails closed, same environment configuration.
    """

    def __init__(self, endpoint: str, api_key: str, timeout: float = 3.0):
        self._sync = PolicyEnforcer(endpoint, api_key, timeout)

    @classmethod
    def from_env(cls) -> "AsyncPolicyEnforcer | None":
        sync = PolicyEnforcer.from_env()
        if sync is None:
            return None
        return cls(sync.base, sync.api_key, sync.timeout)

    @property
    def base(self) -> str:
        return self._sync.base

    @property
    def api_key(self) -> str:
        return self._sync.api_key

    @property
    def timeout(self) -> float:
        return self._sync.timeout

    @property
    def gate_timeout(self) -> float:
        return self._sync.gate_timeout

    async def decide(
        self,
        tool_name: str,
        tool_args: dict,
        *,
        agent_id: str = "",
        invocation_id: str = "",
        request_id: str = "",
    ) -> dict:
        return await asyncio.to_thread(
            self._sync.decide,
            tool_name,
            tool_args,
            agent_id=agent_id,
            invocation_id=invocation_id,
            request_id=request_id,
        )

    async def await_gate(self, approval: dict, *, timeout: float | None = None) -> str:
        status_url = (approval or {}).get("status_url")
        if not status_url:
            return "timeout"
        poll_url = self._sync.base + status_url
        deadline = time.monotonic() + (
            self._sync.gate_timeout if timeout is None else timeout
        )
        while True:
            try:
                status = await asyncio.to_thread(self._sync._poll_gate_once, poll_url)
            except Exception as e:
                logger.warning("gate poll failed (%s); holding the gate closed", e)
                return "timeout"
            if status != "pending":
                return status
            if time.monotonic() >= deadline:
                return "timeout"
            await asyncio.sleep(GATE_POLL_INTERVAL_SECONDS)


def _verdict_message(verdict: dict) -> str:
    decision = verdict.get("decision", "")
    rule = verdict.get("rule", "")
    return verdict.get("message") or f"{decision} by policy {rule}".strip()


def enforce(
    enforcer,
    tool_name: str,
    tool_args: dict,
    *,
    agent_id: str = "",
    invocation_id: str = "",
    request_id: str = "",
) -> "tuple[bool, str]":
    """One decide-and-gate cycle: (allowed, cancel_message).

    ALLOW and an approved GATE return (True, ""); BLOCK and a denied, expired,
    or unanswered GATE return (False, message) where the message is what the
    model should see in place of the tool result. This is the framework-neutral
    core the adapters share; enforcer is any object with the PolicyEnforcer
    decide/await_gate surface.
    """
    verdict = enforcer.decide(
        tool_name,
        tool_args,
        agent_id=agent_id,
        invocation_id=invocation_id,
        request_id=request_id,
    )
    decision = verdict.get("decision")
    if decision == "BLOCK":
        return False, _verdict_message(verdict)
    if decision == "GATE":
        status = enforcer.await_gate(verdict.get("approval") or {})
        if status == "approved":
            return True, ""
        return False, f"{_verdict_message(verdict)} (approval {status})"
    return True, ""


async def enforce_async(
    enforcer,
    tool_name: str,
    tool_args: dict,
    *,
    agent_id: str = "",
    invocation_id: str = "",
    request_id: str = "",
) -> "tuple[bool, str]":
    """enforce() for an AsyncPolicyEnforcer-shaped enforcer."""
    verdict = await enforcer.decide(
        tool_name,
        tool_args,
        agent_id=agent_id,
        invocation_id=invocation_id,
        request_id=request_id,
    )
    decision = verdict.get("decision")
    if decision == "BLOCK":
        return False, _verdict_message(verdict)
    if decision == "GATE":
        status = await enforcer.await_gate(verdict.get("approval") or {})
        if status == "approved":
            return True, ""
        return False, f"{_verdict_message(verdict)} (approval {status})"
    return True, ""


def ambient_otel_trace_id() -> str:
    """The current OpenTelemetry trace id as 32-char hex, "" when there is no
    recording span or no opentelemetry install.

    Sent as invocation_id so the backend can join the enforcement record to the
    trace an OTel-instrumented framework run is already exporting.
    """
    try:
        from opentelemetry import trace

        ctx = trace.get_current_span().get_span_context()
        if not ctx.is_valid:
            return ""
        return format(ctx.trace_id, "032x")
    except Exception:
        return ""

File: js/gentrail-ai/src/enforcement.ts
import { randomUUID } from "node:crypto";

import { trace } from "@opentelemetry/api";

export interface Approval {
  status_url?: string;
}

export interface Verdict {
  decision?: string;
  rule?: string;
  message?: string;
  approval?: Approval;
}

export interface PolicyEnforcerOptions {
  endpoint: string;
  apiKey: string;
  decideTimeoutMs?: number;
  gateTimeoutMs?: number;
  gatePollIntervalMs?: number;
}

export interface DecideOptions {
  agentId?: string;
  invocationId?: string;
  requestId?: string;
}

export interface EnforcementOutcome {
  allowed: boolean;
  message: string;
  decision: string;
  rule: string;
}

const USER_AGENT = "gentrail-sdk-js";
const DECIDE_TIMEOUT_MS_DEFAULT = 3000;
const GATE_TIMEOUT_SECONDS_DEFAULT = 120;
const GATE_POLL_INTERVAL_MS_DEFAULT = 2000;
const INVALID_TRACE_ID = "00000000000000000000000000000000";

export class PolicyEnforcer {
  readonly endpoint: string;
  readonly apiKey: string;
  readonly decideTimeoutMs: number;
  readonly gateTimeoutMs: number;
  readonly gatePollIntervalMs: number;

  constructor(options: PolicyEnforcerOptions) {
    const endpoint = options.endpoint.trim().replace(/\/+$/, "");
    const apiKey = options.apiKey.trim();
    if (endpoint === "") {
      throw new Error("PolicyEnforcer needs a decide endpoint");
    }
    if (apiKey === "") {
      throw new Error("PolicyEnforcer needs an API key");
    }
    this.endpoint = endpoint;
    this.apiKey = apiKey;
    this.decideTimeoutMs = positiveOrDefault(options.decideTimeoutMs, DECIDE_TIMEOUT_MS_DEFAULT);
    this.gateTimeoutMs = options.gateTimeoutMs ?? gateTimeoutMsFromEnv();
    this.gatePollIntervalMs = positiveOrDefault(
      options.gatePollIntervalMs,
      GATE_POLL_INTERVAL_MS_DEFAULT,
    );
  }

  static fromEnv(): PolicyEnforcer | null {
    const endpoint = (process.env.GENTRAIL_DECIDE_ENDPOINT ?? "").trim();
    const apiKey = (process.env.GENTRAIL_API_KEY ?? "").trim();
    if (endpoint === "" || apiKey === "") {
      return null;
    }
    return new PolicyEnforcer({ endpoint, apiKey });
  }

  async decide(toolName: string, toolArgs: unknown, options: DecideOptions = {}): Promise<Verdict> {
    const payload: Record<string, unknown> = {
      event_type: "tool_call",
      tool_name: toolName,
      tool_args: toolArgs ?? {},
      request_id: options.requestId || randomUUID(),
    };
    if (options.agentId) {
      payload.agent_id = options.agentId;
    }
    const invocationId = options.invocationId || ambientOtelTraceId();
    if (invocationId) {
      payload.invocation_id = invocationId;
    }
    try {
      return await this.postDecide(payload);
    } catch (err) {
      console.warn(`gentrail decide failed (${String(err)}); allowing tool ${toolName}`);
      return { decision: "ALLOW" };
    }
  }

  async awaitGate(approval: Approval | undefined): Promise<string> {
    const statusUrl = approval?.status_url;
    if (!statusUrl) {
      return "timeout";
    }
    const pollUrl = this.endpoint + statusUrl;
    const deadline = performance.now() + this.gateTimeoutMs;
    for (;;) {
      let status: string;
      try {
        status = await this.pollGateOnce(pollUrl);
      } catch (err) {
        console.warn(`gentrail gate poll failed (${String(err)}); holding the gate closed`);
        return "timeout";
      }
      if (status !== "pending") {
        return status;
      }
      if (performance.now() >= deadline) {
        return "timeout";
      }
      await sleep(this.gatePollIntervalMs);
    }
  }

  async enforce(
    toolName: string,
    toolArgs: unknown,
    options: DecideOptions = {},
  ): Promise<EnforcementOutcome> {
    const verdict = await this.decide(toolName, toolArgs, options);
    const decision = verdict.decision ?? "";
    const rule = verdict.rule ?? "";
    if (decision === "BLOCK") {
      return { allowed: false, message: verdictMessage(verdict), decision, rule };
    }
    if (decision === "GATE") {
      const status = await this.awaitGate(verdict.approval);
      if (status === "approved") {
        return { allowed: true, message: "", decision, rule };
      }
      return {
        allowed: false,
        message: `${verdictMessage(verdict)} (approval ${status})`,
        decision,
        rule,
      };
    }
    return { allowed: true, message: "", decision, rule };
  }

  private async postDecide(payload: Record<string, unknown>): Promise<Verdict> {
    const response = await fetch(`${this.endpoint}/api/v1/decide`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.apiKey}`,
        "User-Agent": USER_AGENT,
      },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(this.decideTimeoutMs),
    });
    if (!response.ok) {
      throw new Error(`decide returned ${response.status}`);
    }
    return parseVerdict(await response.json());
  }

  private async pollGateOnce(pollUrl: string): Promise<string> {
    const response = await fetch(pollUrl, {
      headers: { Authorization: `Bearer ${this.apiKey}`, "User-Agent": USER_AGENT },
      signal: AbortSignal.timeout(this.decideTimeoutMs),
    });
    if (!response.ok) {
      throw new Error(`gate status returned ${response.status}`);
    }
    const body: unknown = await response.json();
    if (!isRecord(body)) {
      throw new Error("gate status is not a JSON object");
    }
    if (body.status === undefined) {
      return "pending";
    }
    if (typeof body.status !== "string") {
      throw new Error("gate status is not a string");
    }
    return body.status;
  }
}

function parseVerdict(body: unknown): Verdict {
  if (!isRecord(body)) {
    throw new Error("decide verdict is not a JSON object");
  }
  const verdict: Verdict = {};
  if (typeof body.decision === "string") {
    verdict.decision = body.decision;
  }
  if (typeof body.rule === "string") {
    verdict.rule = body.rule;
  }
  if (typeof body.message === "string") {
    verdict.message = body.message;
  }
  if (isRecord(body.approval)) {
    verdict.approval =
      typeof body.approval.status_url === "string" ? { status_url: body.approval.status_url } : {};
  }
  return verdict;
}

export function verdictMessage(verdict: Verdict): string {
  return verdict.message || `${verdict.decision ?? ""} by policy ${verdict.rule ?? ""}`.trim();
}

export function ambientOtelTraceId(): string {
  const traceId = trace.getActiveSpan()?.spanContext().traceId;
  if (!traceId || traceId === INVALID_TRACE_ID) {
    return "";
  }
  return traceId;
}

function gateTimeoutMsFromEnv(): number {
  const seconds = Number(
    process.env.GENTRAIL_GATE_TIMEOUT_SECONDS ?? String(GATE_TIMEOUT_SECONDS_DEFAULT),
  );
  if (!Number.isFinite(seconds) || seconds < 0) {
    return GATE_TIMEOUT_SECONDS_DEFAULT * 1000;
  }
  return seconds * 1000;
}

function positiveOrDefault(value: number | undefined, fallback: number): number {
  if (value === undefined) {
    return fallback;
  }
  if (!Number.isFinite(value) || value <= 0) {
    throw new RangeError(`expected a positive duration, got ${value}`);
  }
  return value;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function sleep(durationMs: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, durationMs));
}
